hydra-node/src/Hydra/Network/Ouroboros.hs

```diff
@@ -143,6 +143,20 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
       _ -> error "getAdrrInfo failed.. do proper error handling"
 
   connect iomgr newBroadcastChannel app = do
+    -- NOTE: We start two concurrent threads, this one connecting to other peers
+    --       and the listen one, listening for other peers to connect.
+    --       When all nodes start at the same time, to be a good citizen, it
+    --       would be fair to first listen for the other peers to connect
+    --       before trying to connect to the other peers.
```

Comment:
Can't we do this instead?

Looking at the `listen` code, it seems that the continuation there is invoked after starting to listen for connections. Would setting up the `listen`, then start the `connect` threads and only then `wait serverAsync` (see line 217) help this issue?

Anyhow. This is also fine the way it is.

Reply:
Interesting. Let me take a look.

Reply:
Thanks for your observation @ch1bo . I love the new code way better.

However, I'm not sure what are the consequences of having removed this instruction:
`wait serverAsync`

We'll see what the C.I. has to say about it but if you already know the answer, I'm interested.

Reply:
We likely need to pass the `serverAsync` to the continuation and call it still after start `connect`ing.

Reply:
Actually, I don't get what's the purpose of the `wait serverAsync` instruction. Could you explain it?

Reply:
`wait serverAsync` is essential as it ensures that, should the Ouroboros server thread crash for any exception, our code is aware of it and terminates its execution. Without it, we wouldn't notice that a thread has crashed and keep running but in an inconsistent state.